gui/biomes.lua

```diff
@@ -0,0 +1,379 @@
+local PEN_NO_ICON = nil
+
+function BiomeVisualizerLegend:get_icon_pen_callback(ix)
+    return function ()
+        if self.SelectedIndex == ix then
+            return PEN_ACTIVE_ICON
+        else
+            return PEN_NO_ICON
+        end
+    end
+end
+
+function BiomeVisualizerLegend:onSelectEntry(idx, option)
+    self.SelectedIndex = idx
+    self.SelectedOption = option
+
+    self:ShowTooltip(option)
+end
+
+function BiomeVisualizerLegend:UpdateChoices()
+    local choices = self.list:getChoices() or {}
+    for i = #choices + 1, #biomeList do
+        local biomeExt = biomeList[i]
+        table.insert(choices, {
+            text = ([[%s: %s]]):format(biomeExt.char, GetBiomeName(biomeExt.biome, biomeExt.typeId)),
+            icon = self:get_icon_pen_callback(#choices+1),
+            biomeTypeId = biomeExt.typeId,
+            biome = biomeExt.biome,
+        })
+    end
+    self.list:setChoices(choices)
+end
+
+do -- implementation of onMouseHoverEntry(idx, option)
+    function BiomeVisualizerLegend:onRenderFrame(dc, rect)
+        BiomeVisualizerLegend.super.onRenderFrame(self, dc, rect)
+
+        local list = self.list
+        local currentHoverIx = list:getIdxUnderMouse()
+        local oldIx = self.HoverIndex
+        if currentHoverIx ~= oldIx then
+            self.HoverIndex = currentHoverIx
+            if self.onMouseHoverEntry then
+                local choices = list:getChoices()
+                self:onMouseHoverEntry(currentHoverIx, choices[currentHoverIx])
+            end
+        end
+    end
+end
+
+local function add_field_text(lines, biome, field_name)
+    lines[#lines+1] = ("%s: %s"):format(field_name, biome[field_name])
+    lines[#lines+1] = NEWLINE
+end
+
+local function get_tooltip_text(option)
+    local text = {}
+    text[#text+1] = ("type: %s"):format(df.biome_type[option.biomeTypeId])
+    text[#text+1] = NEWLINE
+
+    local biome = option.biome
+
+    add_field_text(text, biome, "savagery")
+    add_field_text(text, biome, "evilness")
+    table.insert(text, NEWLINE)
+
+    add_field_text(text, biome, "elevation")
+    add_field_text(text, biome, "rainfall")
+    add_field_text(text, biome, "drainage")
+    add_field_text(text, biome, "vegetation")
+    add_field_text(text, biome, "temperature")
+    add_field_text(text, biome, "volcanism")
+    table.insert(text, NEWLINE)
+
+    local flags = biome.flags
+    if flags.is_lake then
+        text[#text+1] = "lake"
+        text[#text+1] = NEWLINE
+    end
+    if flags.is_brook then
+        text[#text+1] = "brook"
+        text[#text+1] = NEWLINE
+    end
+
+    return text
+end
+
+function BiomeVisualizerLegend:onMouseHoverEntry(idx, option)
+    self:ShowTooltip(option or self.SelectedOption)
+end
+
+function BiomeVisualizerLegend:ShowTooltip(option)
+    local text = get_tooltip_text(option)
+
+    local tooltip_panel = self.tooltip_panel
+    local lbl = tooltip_panel.subviews.label
+
+    lbl:setText(text)
```

Comment:
```suggestion
    lbl:setText(text)
    lbl:updateLayout()
```

Reply:
this causes

```
...steamapps\common\Dwarf Fortress\hack\lua\gui\widgets.lua:51: attempt to index a nil value (local 'parent_rect')
stack traceback:
        ...steamapps\common\Dwarf Fortress\hack\lua\gui\widgets.lua:51: in method 'computeFrame'
        ...Library\steamapps\common\Dwarf Fortress\hack\lua\gui.lua:514: in method 'updateLayout'
        ...amapps\common\Dwarf Fortress/hack/scripts/gui/biomes.lua:323: in method 'ShowTooltip'
```

Also, for me, the info is shown on the initial display.

Reply:
I think this could be fixed if you use a WrappedLabel and set `text_to_wrap=function() return get_tooltip_text(self:get_option()) end,`

Reply:
could you please check if this is needed at all? I couldn't reproduce the initial issue this should fix: the tooltip is there from the start and is updated on hover

Reply:
latest code shows this is fixed.